compiler/packages/babel-plugin-react-compiler/src/HIR/CollectHoistablePropertyLoads.ts

```diff
@@ -0,0 +1,517 @@
+import {CompilerError} from '../CompilerError';
+import {inRange} from '../ReactiveScopes/InferReactiveScopeVariables';
+import {Set_intersect, Set_union, getOrInsertDefault} from '../Utils/utils';
+import {
+  BasicBlock,
+  BlockId,
+  DeclarationId,
+  GeneratedSource,
+  HIRFunction,
+  Identifier,
+  Place,
+  PropertyLoad,
+  ReactiveScopeDependency,
+  ScopeId,
+  TInstruction,
+} from './HIR';
+
+export type CollectHoistablePropertyLoadsResult = {
+  nodes: Map<ScopeId, BlockInfo>;
+  temporaries: Map<Identifier, Identifier>;
+  properties: Map<Identifier, ReactiveScopeDependency>;
+};
+
+/**
+ * Helper function for `PropagateScopeDependencies`.
+ * Uses control flow graph analysis to determine which `Identifier`s can
+ * be assumed to be non-null objects, on a per-block basis.
+ *
+ * Here is an example:
+ * ```js
+ * function useFoo(x, y, z) {
+ *   // NOT safe to hoist PropertyLoads here
+ *   if (...) {
+ *     // safe to hoist loads from x
+ *     read(x.a);
+ *     return;
+ *   }
+ *   // safe to hoist loads from y, z
+ *   read(y.b);
+ *   if (...) {
+ *     // safe to hoist loads from y, z
+ *     read(z.a);
+ *   } else {
+ *     // safe to hoist loads from y, z
+ *     read(z.b);
+ *   }
+ *   // safe to hoist loads from y, z
+ *   return;
+ * }
+ * ```
+ *
+ * Note that we currently do NOT account for mutable / declaration range
+ * when doing the CFG-based traversal, producing results that are technically
+ * incorrect but filtered by PropagateScopeDeps (which only takes dependencies
+ * on constructed value -- i.e. a scope's dependencies must have mutable ranges
+ * ending earlier than the scope start).
+ *
+ * Take this example, this function will infer x.foo.bar as non-nullable for bb0,
+ * via the intersection of bb1 & bb2 which in turn comes from bb3. This is technically
+ * incorrect bb0 is before / during x's mutable range.
+ *  bb0:
+ *    const x = ...;
+ *    if cond then bb1 else bb2
+ *  bb1:
+ *    ...
+ *    goto bb3
+ *  bb2:
+ *    ...
+ *    goto bb3:
+ *  bb3:
+ *    x.foo.bar
+ */
+export function collectHoistablePropertyLoads(
+  fn: HIRFunction,
+  usedOutsideDeclaringScope: Set<DeclarationId>,
+): CollectHoistablePropertyLoadsResult {
+  const sidemap = collectSidemap(fn, usedOutsideDeclaringScope);
+  const nodes = collectNodes(fn, sidemap);
+  deriveNonNull(fn, nodes);
+
+  const nodesKeyedByScopeId = new Map<ScopeId, BlockInfo>();
+  for (const [_, block] of fn.body.blocks) {
+    if (block.terminal.kind === 'scope') {
+      nodesKeyedByScopeId.set(
+        block.terminal.scope.id,
+        nodes.get(block.terminal.block)!,
+      );
+    }
+  }
+
+  return {
+    nodes: nodesKeyedByScopeId,
+    temporaries: sidemap.temporaries,
+    properties: sidemap.properties,
+  };
+}
+
+export type BlockInfo = {
+  block: BasicBlock;
+  assumedNonNullObjects: Set<PropertyLoadNode>;
+};
+
+export function getProperty(
```

Comment:
`getProperty` and `resolveTemporary` were essentially moved from PropagateScopeDeps (reactive scopes version). Understanding hoistable property loads required the same `properties` and `temporaries` sidemaps as the core dependency-extraction logic. Now, we first compute the sidemaps, then use them (as `ReadOnlyMaps`) in both `CollectHoistablePropertyLoads` and `PropagateScopeDeps`